include/flux/op/set_algorithms.hpp

```diff
@@ -0,0 +1,576 @@
+
+            friend auto operator==(cursor_type const&, cursor_type const&) -> bool
+                requires std::equality_comparable<cursor_t<Base1>> &&
+                         std::equality_comparable<cursor_t<Base2>>
+                = default;
+        };
+
+        template <typename Self>
+        static inline constexpr bool maybe_const_iterable
+            = std::is_const_v<Self> ? flux::sequence<Base1 const> : true;
+
+        template <typename Self>
+        static constexpr void update(Self& self, cursor_type& cur) {
+            while(not flux::is_last(self.base1_, cur.base1_cursor))
+            {
+                if(flux::is_last(self.base2_, cur.base2_cursor)) {
+                    return;
+                }
+
+                if(std::invoke(self.cmp_, flux::read_at(self.base1_, cur.base1_cursor), 
+                                          flux::read_at(self.base2_, cur.base2_cursor))) {
+                    return;
+                }
+
+                if(not std::invoke(self.cmp_, flux::read_at(self.base2_, cur.base2_cursor), 
+                                              flux::read_at(self.base1_, cur.base1_cursor))) {
+                    flux::inc(self.base1_, cur.base1_cursor);
+                }
+
+                flux::inc(self.base2_, cur.base2_cursor);
+            }
+        }
+
+    public:
+        using value_type = value_t<Base1>;
+
+        inline static constexpr bool is_infinite = flux::infinite_sequence<Base1>;
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto first(Self& self) -> cursor_type
+        {
+            auto cur = cursor_type{.base1_cursor = flux::first(self.base1_),
+                                   .base2_cursor = flux::first(self.base2_)};
+            update(self, cur);
+            return cur;
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto is_last(Self& self, cursor_type const& cur) -> bool
+        {
+            return flux::is_last(self.base1_, cur.base1_cursor);
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto inc(Self& self, cursor_type& cur) -> void
+        {
+            flux::inc(self.base1_, cur.base1_cursor);
+            update(self, cur);
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self>
+        static constexpr auto read_at(Self& self, cursor_type const& cur)
+            -> decltype(flux::read_at(self.base1_, cur.base1_cursor))
+        {
+            return flux::read_at(self.base1_, cur.base1_cursor);
+        }
+
+        template <typename Self>
+            requires maybe_const_iterable<Self> && bounded_sequence<Base1>
+        static constexpr auto last(Self& self) -> cursor_type
+        {
+            return cursor_type{flux::last(self.base1_), flux::last(self.base2_)};
+        }
```

Comment:
This function is calling `flux::last(self.base2_)`, so it should make sure `Base2` is also a `bounded_sequence`.

BUT...

I'm wondering what happens when `seq1` contains fewer elements than `seq2`, and we iterate over `set_difference(seq1, seq2)` until we get to the end... in that case the cursor will contain `(base1_cur, base2_cur)` where `is_last(seq1, base_cur1)` is true, but `is_last(seq2, base_cur2)` is not true.

This cursor then ends up not comparing equal to the cursor which would be returned from `last()`:

```cpp
auto seq1 = std::vector{1, 3, 5};
auto seq2 = std::vector{2, 4, 6, 7, 10};

auto diff = set_difference(ref(seq1), ref(seq2));

// Iterate to the end
auto cur = diff.find_if(flux::pred::false_);

// cur is now definitely an end cursor...
assert(diff.is_last(cur));

// ...but it's not equal to last()?!
assert(cur == diff.last()); // fails
```

This is a serious problem, and I don't know what right solution is :(

For now, I think it's best if we say that `set_difference_adaptor` a never bounded sequence -- that is, remove the `last()` function.